pkg/apis/core/validation/validation.go

```diff
@@ -3075,6 +3075,39 @@ func validateContainerCommon(ctr *core.Container, volumes map[string]core.Volume
 	allErrs = append(allErrs, validatePullPolicy(ctr.ImagePullPolicy, path.Child("imagePullPolicy"))...)
 	allErrs = append(allErrs, ValidateResourceRequirements(&ctr.Resources, path.Child("resources"), opts)...)
 	allErrs = append(allErrs, ValidateSecurityContext(ctr.SecurityContext, path.Child("securityContext"))...)
+	return allErrs
+}
+
+func validateHostUsers(spec *core.PodSpec, volumes map[string]core.VolumeSource, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport) {
+		if spec.SecurityContext != nil && spec.SecurityContext.HostUsers != nil {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("hostUsers"), "field not supported. Enable the feature gate UserNamespacesSupport"))
+		}
+		return allErrs
+	}
+
+	// Only make the following checks if hostUsers is false (otherwise, the container uses the
+	// same userns as the host, and so there isn't anything to check).
+	if spec.SecurityContext == nil || spec.SecurityContext.HostUsers == nil || *spec.SecurityContext.HostUsers == true {
+		return allErrs
+	}
+
+	// For now only these volumes are supported:
+	// - configmap
+	// - secret
+	// - downwardAPI
+	// - emptyDir
+	// - projected
+	// So reject anything else.
```

Comment:
The [graduation criteria](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/127-user-namespaces#graduation-criteria) doesn't make it clear what the plan for this restriction is. Are other volume types (especially persistent volumes) a graduation criteria before beta? Is that going to change the shape of the API?

I'm not used to seeing API changes get added/released with that level of ambiguity in the alpha → beta graduation criteria, and no GA graduation criteria.

Reply:
It looks like the KEP does have a phase 2 for how to handle the remaining volume types, although it is not mentioned in graduation criteria. https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/127-user-namespaces#phase-2-pods-with-volumes.

It would be good to have sig-storage as a participating reviewer on the KEP for future phases.

Reply:
Exactly, in later phases (phase II and phase III), volume support will be added and those are a pre-requisite for beta, we talked that with Mruanl :)

We haven't put that in the KEP, as those phases are not completely defined yet, so to not add unclear criteria.

Reply:
I see a phase 2 mentioned in the KEP, but not in the graduation criteria, and a phase 3 with lots of TBD. I'm finding it hard to understand the enablement/compat implications of the field that's being added in this PR, which I suspect comes largely from me lacking context on the design. @thockin is mentioned in the [phase 3](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/127-user-namespaces#phase-3-tbd) bits a lot ... I wonder if he would be a better API reviewer for this, knowing where the feature is going.

Reply:
@derekwaynecarr yes, exactly :)

@liggitt if you can elaborate on what you find hard to understand regarding the enablement of this feature, I can try to explain (and open PRs with clarifications in the KEP). So far we have clarified what was mentioned by reviewers (PRR review, etc.) but your concerns have not come up yet :)

Alternatively, if @thockin wants to review that would be great too :)

Reply:
@derekwaynecarr it is not clear because it doesn't happen. In phase 1 we have only volumes that have the same lifecycle of the pod (configmaps, secret, downwardsAPI, projected, emptyDir). So, when we create those volumes we create them with the hostIDs for that pod (see https://github.com/kubernetes/kubernetes/pull/111090/files#diff-3e8f8a6e5eca57db8e84bcc6c896f0a3b10289fe55d8e999b4bdf1cfdc0de394, then configmap, etc are modified to honor that field).

In phase II, when pods have other volume types, we will assign them always the same mapping for those pods (and only those pods). But we are not in phase II, that is why you don't see it :)

Reply:
> @rata said:
>> @liggitt if you can elaborate on what you find hard to understand regarding the enablement of this feature, I can try to explain (and open PRs with clarifications in the KEP).

There were several open questions in the [beta graduation criteria](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/127-user-namespaces#beta) that look like they could impact the shape of the API (Should we allow any way for users to for "more" IDs mapped? If yes, how many more and how? Should we allow the user to ask for specific mappings?)

I expected those to be answered before starting to merge API changes, since the answers to those questions can change the shape of the API.

> @derekwaynecarr said:
>> @rata i think the best clarification is to highlight how a consistent remapping is applied across all nodes in a fleet for pods that use persistent volumes.

This is one of the key pieces I was trying to understand how it would work. I was looking for phase 2 to be in the [beta graduation criteria](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/127-user-namespaces#beta), and phase 2 and phase 3 weren't mentioned at all, so it was difficult to tell if the intent was to enable the user namespacing feature ahead of solving the PV bits.

If this feature can't move to beta until PVs are supported, it's hard to know if the API shape in this PR is a good starting point.

> @rata said:
>> In phase II, when pods have other volume types, we will assign them always the same mapping for those pods (and only those pods)

The KEP for phase 2 says `All pods in this mode will use the same mapping, chosen by the kubelet, with a length 65536, and mapping the range 0-65535 too.` Is there API surface required to make sure all kubelets in the cluster choose the same mapping for persistent volume pods?

Reply:
> There were several open questions in the [beta graduation criteria](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/127-user-namespaces#beta) that look like they could impact the shape of the API (Should we allow any way for users to for "more" IDs mapped? If yes, how many more and how? Should we allow the user to ask for specific mappings?)
> 
> I expected those to be answered before starting to merge API changes, since the answers to those questions can change the shape of the API.

The answer is no, specially now that the scope is just stateless pods (see https://github.com/kubernetes/kubernetes/pull/111090#issuecomment-1203066490). I can update the KEP tomorrow.

> If this feature can't move to beta until PVs are supported, it's hard to know if the API shape in this PR is a good starting point.

Idem, see the comment. We changed the scope to only stateless pods and we will migrate to beta only support for stateless pods.

> The KEP for phase 2 says `All pods in this mode will use the same mapping, chosen by the kubelet, with a length 65536, and mapping the range 0-65535 too.` Is there API surface required to make sure all kubelets in the cluster choose the same mapping for persistent volume pods?

No. But no need to discuss it further, phase II is out of scope. We will tackle stateless pods only :)

Reply:
@liggitt I think this should address all your concerns. Let me know if anything else is missing!

Reply:
I will look at the PR, though it will be superficial wrt implementation details.

Phase 1 is clearly valuable and should be "safe" in pretty much every way.  I still think that mapping 64K for each pod is wasteful, but we should be able to revisit that without breaking users or APIs.  I think I would be OK to let that proceed to GA independent of phase 2 or 3.  It's a big improvement over staus quo.

Phase 2 makes sense from an incremental engineering POV, but as noted in the KEP, I am not sure we should actually launch it without phase 3.  I agreed to leave tha half-defined in KEP because I thought it was important to move ahead on phase 1.

Reply:
scoping to stateless pods does resolve my comments about the uncertain future API additions

Reply:
@liggitt which concern exactly is not addressed? Can you elaborate? Future API additions are not needed for stateless pods. What am I missing?

Reply:
it *does* resolve this thread :) (I marked it resolved as well)